apisix/balancer/roundrobin.lua

```diff
@@ -16,16 +16,55 @@
 --
 
 local roundrobin  = require("resty.roundrobin")
+local core = require("apisix.core")
+local nkeys = core.table.nkeys
+
 
 local _M = {}
 
 
 function _M.new(up_nodes, upstream)
     local picker = roundrobin:new(up_nodes)
+    local nodes_count = nkeys(up_nodes)
     return {
         upstream = upstream,
-        get = function ()
-            return picker:find()
+        get = function (ctx)
+            if ctx.balancer_tried_servers and nkeys(ctx.balancer_tried_servers) == nodes_count then
+                return nil, "all upstream servers tried"
+            end
+
+            local server, err
+            while true do
+                server, err = picker:find()
+                if not server then
+                    return nil, err
+                end
+                if ctx.balancer_tried_servers then
+                    if not ctx.balancer_tried_servers[server] then
+                        break
+                    end
+                else
+                    break
+                end
+            end
+
+            return server
+        end,
+        after_balance = function (ctx, before_retry)
```

Comment:
It seems the `before_retry` is always `true` when `after_balance` is called?

Reply:
There is another call in `apisix/init.lua`.

Reply:
got it.